apps/history/declarativetabmodel.cpp

```diff
@@ -454,6 +454,20 @@ void DeclarativeTabModel::onUrlChanged()
     }
 }
 
+void DeclarativeTabModel::onDesktopModeChanged()
+{
+    DeclarativeWebPage *webPage = qobject_cast<DeclarativeWebPage *>(sender());
+    if (webPage) {
+        int tabIndex = findTabIndex(webPage->tabId());
+        if (tabIndex >= 0 && m_tabs.at(tabIndex).desktopMode() != webPage->desktopMode()) {
+            QVector<int> roles;
+            roles << UrlRole;
```

Comment:
This is not yet changing url as there will be another load when the mode changes. If url did change due to desktop mode change, then that will be handled onUrlChanged slot.

As you are introducing already desktop mode to the tab, let's add DesktopModeRole role (exposed as "desktopMode"). You need to fix/update data method to return the desktop mode as well.